basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/RegexNodeType.java

```diff
@@ -0,0 +1,97 @@
+package club.mondaylunch.gatos.basicnodes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import club.mondaylunch.gatos.core.data.DataType;
+import club.mondaylunch.gatos.core.graph.connector.NodeConnector;
+import club.mondaylunch.gatos.core.data.DataBox;
+import club.mondaylunch.gatos.core.graph.type.NodeType;
+
+/**
+ * A node that operates on regular expressions.
+ *
+ * <p>
+ * This is a {@link Node node} (addressable by UUID).
+ * </p>
+ * <p>
+ * It takes a regular expression: "regex" and input word: "word".
+ * </p>
+ * <p>
+ * It outputs:
+ * "isMatch" a boolean to confirm if a match was found, if false all other output values are null.
+ * "match" the first matching String.
+ * "group" a List of all groups, if the regex has no groups this returns null.
+ * </p>
+ */
+public class RegexNodeType extends NodeType.Process {
+    @Override
+    public Map<String, DataBox<?>> settings() {
+        return Map.of();
+    }
+
+    @Override
+    public Set<NodeConnector.Input<?>> inputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        return Set.of(
+            new NodeConnector.Input<>(nodeId, "regex", DataType.STRING),
+            new NodeConnector.Input<>(nodeId, "word", DataType.STRING)
+        );
+    }
+
+    @Override
+    public Set<NodeConnector.Output<?>> outputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        return Set.of(
+            new NodeConnector.Output<>(nodeId, "isMatch", DataType.BOOLEAN),
+            new NodeConnector.Output<>(nodeId, "match", DataType.STRING),
+            new NodeConnector.Output<>(nodeId, "group", DataType.STRING.listOf())
+        );
+    }
+
+    @Override
+    public Map<String, CompletableFuture<DataBox<?>>> compute(Map<String, DataBox<?>> inputs, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {        
+        var regex = Pattern.compile(DataBox.get(inputs, "regex", DataType.STRING).orElseThrow());        
+        var word = DataBox.get(inputs, "word", DataType.STRING).orElseThrow();
+        var matcher = regex.matcher(word);
+
+        if (!matcher.find()) return Map.of(
+            "isMatch", CompletableFuture.completedFuture(DataType.BOOLEAN.create(false)),
+            "match", CompletableFuture.completedFuture(DataType.STRING.create(null)),
+            "groups", CompletableFuture.completedFuture(DataType.STRING.listOf().create(null))
```

Comment:
We never return nulls from nodes. If there's a possibility that something could not have an output, use an optional type.

Reply:
![1984](https://user-images.githubusercontent.com/54268387/226699072-cceb8b0b-2acf-4ddc-afcf-c0f5badf9adb.gif)

Reply:
https://user-images.githubusercontent.com/61510644/226700080-30931186-9331-4049-9aef-43323570bfb6.mp4